arangod/Aql/QueryRegistry.cpp

```diff
@@ -182,146 +188,201 @@ void* QueryRegistry::openEngine(EngineId id, EngineType type) {
-      // first, insert a dummy entry for this database
-      m = _queries
-              .emplace(
-                  vocbase,
-                  std::unordered_map<QueryId, std::unique_ptr<QueryInfo>>())
-              .first;
-    }
+  // remove query from the table of running queries
+  vocbaseQueriesIt->second.erase(queryMapIt);
 
-    auto q = m->second.find(id);
-
-    if (q == m->second.end()) {
-      if (errorCode != TRI_ERROR_NO_ERROR &&
-          errorCode != TRI_ERROR_SHUTTING_DOWN) {
-        // insert a tombstone with a higher-than-query timeout
-        LOG_TOPIC("779f5", DEBUG, arangodb::Logger::AQL)
-            << "inserting tombstone for query " << id << " into query registry";
-        auto inserted =
-            m->second
-                .emplace(id,
-                         std::make_unique<QueryInfo>(
-                             errorCode, std::max(_defaultTTL, 600.0) + 300.0))
-                .second;
-        TRI_ASSERT(inserted);
-        return nullptr;
-      }
+  // remove engines
+  for (auto const& engine : queryInfo->_query->snippets()) {
+#ifndef ARANGODB_ENABLE_MAINTAINER_MODE
+    _engines.erase(engine->engineId());
+#else
 
-      THROW_ARANGO_EXCEPTION_MESSAGE(
-          TRI_ERROR_BAD_PARAMETER,
-          "query with given id not found in query registry");
+    auto it = _engines.find(engine->engineId());
+    if (it != _engines.end()) {
+      TRI_ASSERT(it->second._queryInfo != nullptr);
+      TRI_ASSERT(!it->second._isOpen);
+      _engines.erase(it);
     }
+#endif
+  }
+  for (auto& engine : queryInfo->_query->traversers()) {
+    _engines.erase(engine->engineId());
+  }
 
-    if (q->second->_isTombstone) {
-      m->second.erase(q);
-      return nullptr;
-    }
+  if (vocbaseQueriesIt->second.empty()) {
+    // clear empty entries in database-to-queries map
+    _queries.erase(vocbaseQueriesIt);
+  }
 
-    if (q->second->_numOpen > 0) {
-      TRI_ASSERT(!q->second->_isTombstone);
+  LOG_TOPIC("6756c", DEBUG, arangodb::Logger::AQL)
+      << "query with id " << id << " is now destroyed";
+}
 
-      // query in use by another thread/request
-      if (errorCode == TRI_ERROR_QUERY_KILLED) {
-        q->second->_query->kill();
-      }
-      q->second->_expires = 0.0;
-      return nullptr;
-    }
+futures::Future<std::shared_ptr<ClusterQuery>> QueryRegistry::finishQuery(
+    std::string const& vocbase, QueryId id, ErrorCode errorCode) {
+  WRITE_LOCKER(writeLocker, _lock);
 
-    // move query into our unique ptr, so we can process it outside
-    // of the lock
-    queryInfo = std::move(q->second);
+  QueryInfoMapPerVocbase::iterator vocbaseQueriesIt;
+  QueryInfoMap::iterator queryMapIt;
+  if (!lookupQueryForFinalization(vocbase, id, errorCode, vocbaseQueriesIt,
+                                  queryMapIt)) {
+    return std::shared_ptr<ClusterQuery>();
+  }
 
-    // remove query from the table of running queries
-    m->second.erase(q);
+  TRI_ASSERT(queryMapIt->second);
```

Comment:
I think we should add assertions here that both `vocbaseQueriesIt` and `queryMapIt` are populated as expected and are valid iterators.